src/Native/Runtime/arm/PInvoke.S

```diff
@@ -3,39 +3,74 @@
 // See the LICENSE file in the project root for more information.
 
 #include <unixasmmacros.inc>
+#include <AsmOffsets.inc>
 
 .syntax unified
 .thumb
 
-#ifdef _DEBUG
-.rodata
-AssertMsg:                              .asciz "__FILE__:%s: %s is not implemented\n"
-FileName:                               .asciz "PInvoke.S"
-RhpPInvokeName:                         .asciz "RhpPInvoke"
-RhpPInvokeReturnName:                   .asciz "RhpPInvokeReturn"
+.macro INLINE_GET_TLS_VAR Var, trashReg
+	        ldr	         r0, =\Var(tlsgd)
+	        bl	         __tls_get_addr@PLT
+	        ldr	         \trashReg, =\Var@DTPOFF
+	        add	         r0, \trashReg
+.endmacro
 
-.text
-.macro GEN_ASSERT_FUNC func
-        GEN_ASSERT AssertMsg, FileName, \func
-.endm
-#endif
-
-// TODO: Implement Arm support
+.macro INLINE_GETTHREAD trashReg
+	        // Inlined version of call C_FUNC(RhpGetThread)
+	        INLINE_GET_TLS_VAR tls_CurrentThread, \trashReg
+.endmacro
 
 //
 // RhpPInvoke
 //
+// IN:  R0: address of pinvoke frame
+//
+// This helper assumes that its callsite is as good to start the stackwalk as the actual PInvoke callsite.
+// The codegenerator must treat the callsite of this helper as GC triggering and generate the GC info for it.
+// Also, the codegenerator must ensure that there are no live GC references in callee saved registers.
+//
 NESTED_ENTRY RhpPInvoke, _TEXT, NoHandler
-#ifdef _DEBUG
-        GEN_ASSERT_FUNC RhpPInvokeName
-#endif
+	        PROLOG_PUSH "{r4-r7,lr}"
+
+	        str         lr, [r0, #OFFSETOF__PInvokeTransitionFrame__m_RIP]
+	        str         r7, [r0, #OFFSETOF__PInvokeTransitionFrame__m_FramePointer]
+	        str         sp, [r0, #OFFSETOF__PInvokeTransitionFrame__m_PreservedRegs]
+
+	        mov	        r5, r0
+	        // R0 = GetThread(), R6 - Trash Register
+	        INLINE_GETTHREAD r6
+	        str         r0, [r5, #OFFSETOF__PInvokeTransitionFrame__m_pThread]
+	        str         r5, [r0, #OFFSETOF__Thread__m_pTransitionFrame]
+
+	        ldr         r6, =C_FUNC(RhpTrapThreads)
+	        ldr         r6,	[r6]
+	        cbnz        r6, LOCAL_LABEL(InvokeRareTrapThread)
+
+	        EPILOG_POP  "{r4-r7,lr}"
```

Comment:
It is more common to do `EPILOG_POP  "{r4-r7,pc}"` here, and then you do not need the `bx lr` on the next line.